apps/hash-frontend/src/pages/shared/use-kratos-flow-error-handler.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  ErrorAuthenticatorAssuranceLevelNotSatisfied,
+  ErrorBrowserLocationChangeRequired,
+  NeedsPrivilegedSessionError,
+} from "@ory/client";
+import { AxiosError } from "axios";
+import { useRouter } from "next/router";
+import { Dispatch, SetStateAction, useCallback } from "react";
+
+import { useAuthInfo } from "./auth-info-context";
+import { Flows } from "./ory-kratos";
+
+export const useKratosErrorHandler = <S>(props: {
+  flowType: keyof Flows;
+  setFlow: Dispatch<SetStateAction<S | undefined>>;
+  setErrorMessage: Dispatch<SetStateAction<string | undefined>>;
+}) => {
+  const { flowType, setFlow, setErrorMessage } = props;
+
+  const { authenticatedUser } = useAuthInfo();
+  const router = useRouter();
+
+  const handleFlowError = useCallback(
+    async (err: AxiosError<any>) => {
+      const kratosError = err.response?.data;
+
+      if (kratosError) {
+        switch (kratosError.error?.id) {
+          case "session_aal2_required": {
+            // 2FA is enabled and enforced, but user did not perform 2FA yet!
+            const { redirect_browser_to } =
+              kratosError as ErrorAuthenticatorAssuranceLevelNotSatisfied;
+
+            if (redirect_browser_to) {
+              await router.replace(redirect_browser_to);
+            }
+            return;
+          }
+          case "session_already_available":
+            // User is already signed in, if we're in the login flow let's redirect them home!
+            if (flowType === "login") {
+              if (!authenticatedUser) {
+                throw new Error(
+                  "The user is authenticated, but the user entity could not be fetched. Check your connection to the backend API, and ensure it is running.",
+                );
+              }
+              await router.push("/");
```

Comment:
Just for my own understanding, is this the main check that prevents the infinite login redirect?

Reply:
Correct, essentially the loop that was occurring was:
1. Navigation to `/login` page
2. Existing kratos session can be used for authentication, so redirect to homepage
3. Navigation to `/`
4. Authenticated user could not be fetcehd, redirect to `/login` page
5. Navigation to `/login` page

And now the loop is broken because we only redirect from the `/login` page to `/` when the authenticated user has been successfully fetched.